### ISSUE TITLE
Clang crashes on ill-formed template code

### ISSUE THREAD

**Issue by Weverything** [llvm/llvm-project#20852]
|  |  |
| --- | --- |
| Bugzilla Link | [20478](https://llvm.org/bz20478) |
| Version | trunk |
| OS | Linux |

## Extended Description 
$ cat var.cc
```cpp
template<class T>
struct Foo {};

struct S {
  static Foo<Bar<int>::type>> test();
};
```
$clang var.cc
var.cc:5:14: error: use of undeclared identifier 'Bar'
  static Foo<Bar<int>::type>> test();
             ^
var.cc:5:21: error: expected '(' for function-style cast or type construction
  static Foo<Bar<int>::type>> test();
                 ~~~^
var.cc:5:24: error: non-friend class member 'type' cannot have a qualified name
  static Foo<Bar<int>::type>> test();
                     ~~^
var.cc:5:24: error: no member named 'type' in the global namespace
  static Foo<Bar<int>::type>> test();
                     ~~^
0  clang-3.5       0x00000000010e6fd5 llvm::sys::PrintStackTrace(_IO_FILE*) + 37
1  clang-3.5       0x00000000010e7423
2  libpthread.so.0 0x00007fa6d8f86cb0
3  clang-3.5       0x0000000001b3a1dc clang::Sema::FinalizeDeclaration(clang::Decl*) + 204
4  clang-3.5       0x0000000001928ee0 clang::Parser::ParseCXXClassMemberDeclaration(clang::AccessSpecifier, clang::AttributeList*, clang::Parser::ParsedTemplateInfo const&, clang::ParsingDeclRAIIObject*) + 7376
5  clang-3.5       0x00000000019256b0 clang::Parser::ParseCXXMemberSpecification(clang::SourceLocation, clang::SourceLocation, clang::Parser::ParsedAttributesWithRange&, unsigned int, clang::Decl*) + 3360
6  clang-3.5       0x0000000001923efd clang::Parser::ParseClassSpecifier(clang::tok::TokenKind, clang::SourceLocation, clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, bool, clang::Parser::DeclSpecContext, clang::Parser::ParsedAttributesWithRange&) + 7853
7  clang-3.5       0x0000000001909db9 clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*) + 1577
8  clang-3.5       0x00000000018fbab0 clang::Parser::ParseDeclOrFunctionDefInternal(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec&, clang::AccessSpecifier) + 64
9  clang-3.5       0x00000000018fb860 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*, clang::AccessSpecifier) + 384
10 clang-3.5       0x00000000018fabec clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) + 3500
11 clang-3.5       0x00000000018f9db4 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) + 388
12 clang-3.5       0x00000000018f5e96 clang::ParseAST(clang::Sema&, bool, bool) + 406
13 clang-3.5       0x0000000001246979 clang::FrontendAction::Execute() + 57
14 clang-3.5       0x000000000121b7c3 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) + 787
15 clang-3.5       0x00000000012b2967 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) + 3207
16 clang-3.5       0x00000000006b3ef0 cc1_main(char const**, char const**, char const*, void*) + 672
17 clang-3.5       0x00000000006b2140 main + 7712
18 libc.so.6       0x00007fa6d84af76d __libc_start_main + 237
19 clang-3.5       0x00000000006b01ed

**Comment by shafik** [comment]
This no longer crashes on trunk: https://godbolt.org/z/P8cd31f3c 

Maybe related to: https://github.com/llvm/llvm-project/issues/24749#event-9531305738

**Comment by Endilll** [comment]
Fixed in Clang 3.6: https://godbolt.org/z/cMPYPj99M
